backend/models/group.js

```diff
@@ -785,46 +776,168 @@ function findObjectsByQuery(account, model, query) {
+ */
+async function getIdToMeshesDict(account, model, revId) {
+	const treeFileName = `${revId}/idToMeshes.json`;
+	const {readStream: fileStream} = await FileRef.getJSONFileStream(account, model, treeFileName);
+
+	const idToMeshes =  await new Promise((resolve, reject) => {
+		let treeStr = "";
+		try {
+			fileStream.on("data", (chunk) => {
+				treeStr += chunk;
+			});
 
-					});
+			fileStream.on("end", () => {
+				try {
+					resolve(JSON.parse(treeStr));
+				} catch(e) {
+					reject(e);
 				}
+
 			});
-		} else {
-			return Promise.resolve([]);
+		} catch(e) {
+			reject(e);
 		}
 	});
+
+	return idToMeshes;
+}
+
+function batchPromises(promiseGenerator, dataToBatch, size) {
+	const promises = [];
+	for(let i = 0; i < dataToBatch.length; i = i + size) {
+		const endIndex = Math.min(i + size , dataToBatch.length);
+		const batchedData = dataToBatch.slice(i, endIndex);
+		promises.push(promiseGenerator(batchedData));
+	}
+	return Promise.all(promises);
+}
+/**
+ *
+ * @param {string} account
+ * @param {string} model
+ * @param {Array<object>} ids
+ * @param {boolean} convertSharedIDsToString
+ *
+ * @returns {Promise<Array<string | object>>}
+ */
+async function idsToSharedIds(account, model, ids, convertSharedIDsToString) {
+	const sceneCol = await db.getCollection(account, model + ".scene");
+
+	const treeItems = await batchPromises((_idsForquery) => {
+		return sceneCol.find({_id: {$in : _idsForquery}}, {shared_id:1 , _id:0}).toArray();
+	}, ids , 7000);
+
+	const shared_ids = [];
+
+	for (let i = 0; i < treeItems.length ; i++) {
+		const treeItemsBatch = treeItems[i];
+		for (let j = 0; j < treeItemsBatch.length ; j++) {
+			const { shared_id } = treeItemsBatch[j];
+			if (convertSharedIDsToString) {
+				shared_ids.push(utils.uuidToString(shared_id));
+			} else {
+				shared_ids.push(shared_id);
+			}
+		}
+	}
+
+	return shared_ids;
+}
+
+/**
+ * Return shared ids resulted of applying all the queries at once
+ * @param {string} account
+ * @param {string} model
+ * @param {Array<object>} ruleQueries
+ * @param {string} branch
+ * @param {string} revId
+ * @param {Boolean} revId
+ *
+ * @returns {Promise<Array<string | object>>}
+ */
+async function findModelSharedIDsByRulesQueries(account, model, ruleQueries, branch, revId, convertSharedIDsToString) {
+	const history = await History.getHistory({ account, model }, branch, revId);
+
+	if (!history) {
+		return Promise.reject(responseCodes.INVALID_TAG_NAME);
+	}
+
+	const idToMeshesDict = await getIdToMeshesDict(account, model, utils.uuidToString(history._id));
+
+	const eachRuleResults = await Promise.all(ruleQueries.map(ruleQuery => getRuleQueryResults(account, model, idToMeshesDict, history.current, ruleQuery)));
+
+	const allRulesResults = multiIntersection(eachRuleResults);
+
+	const ids = [];
+	for (const id of allRulesResults) {
+		ids.push(utils.stringToUUID(id));
+	}
+
+	return await idsToSharedIds(account, model, ids, convertSharedIDsToString) ;
```

Comment:
shared_ids are guaranteed to be unique per object within the same revision.

If you keep all your results in shared_ids instead of _ids, then you wouldn't need this conversion?

Reply:
cant be done